src/cycles/eulerian.jl

```diff
@@ -0,0 +1,315 @@
+# Checking for existence of Eulerian circuit and Eulerian trail
+
+function single_nonzero_degree_component(
+    graph::AG,
+    ) where {T <: Integer, AG <: AbstractGraph{T}}
+
+    visited = zeros(UInt8, nv(graph))
+    for v in vertices(graph)
+        if (outdegree(graph, v) + indegree(graph, v)) == 0
+            visited[v] = 2
+        end
+    end
+    for v in vertices(graph)
+        if (outdegree(graph, v) + indegree(graph, v)) == 0
+            continue
+        end
+        stack = Vector{T}([v])
```

Comment:
You could move the stack out of the foor loop, and then just empty it every in every iteration, so you don't have to allocate it in every iteration.